stellargraph/mapper/adjacency_generators.py

```diff
@@ -0,0 +1,84 @@
+import tensorflow as tf
+from tensorflow.keras import backend as K
+import numpy as np
+from ..core import StellarGraph
+from ..core.utils import normalize_adj
+from ..core.tf_utils import partial_powers, select_row_from_sparse_tensor
+from ..core.experimental import experimental
+
+
+@experimental(reason="lack of unit tests")
+class AdjacencyPowerGenerator:
+    """
+    A data generator for use with the Watch Your Step algorithm [1]. It calculates and returns the first `num_powers`
+    of the adjacency matrix row by row.
+
+    .. warning::
+
+        This class is experimental: it is insufficiently tested.
+
+    Args:
+        G (StellarGraph): a machine-learning StellarGraph-type graph
+        num_powers (int): the number of adjacency powers to calculate
+
+    """
+
+    def __init__(self, G, num_powers=5):
+
+        if not isinstance(G, StellarGraph):
+            raise TypeError("G must be a StellarGraph object.")
+
+        Aadj = G.to_adjacency_matrix()
+        indices = list(zip(Aadj.col, Aadj.row))
```

Comment:
I think this is going to fail if `to_adjacency_matrix` returns something that's not `coo` format, so:

```python
Aadj = G.to_adjacency_matrix().tocoo()
```

Also, doing this processing via python iterators is unnecessarily expensive given these are already numpy arrays:

```suggestion
        indices = np.column_stack((Aadj.col, Aadj.row))
```

Reply:
I didn't know that python iterators were expensive! are they expensive in general or just relative to processing numpy arrays in numpy?

Reply:
Relative to processing numpy arrays in numpy. Python iterators are relatively inexpensive compared to everything else in the rest of Python, but numpy wins because it cuts out all of general Python overhead. In particular, a numpy array is a sequence of small numbers in memory, e.g. `np.arange(10000)` places 8 byte `np.int64` integers contiguously into a 80000 byte chunk of memory. This makes processing very efficient, because CPUs are great at doing operations on big chunks of memory like that. For example:

```python
import numpy as np
data = np.arange(10000)

%timeit -n 100 data.sum()
# 8.67 µs ± 880 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit -n 100 sum(data)
# 1.39 ms ± 94.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# just iterate over 10000 in a Python list
L = [None]*10000
def f():
    for _ in L: pass
%timeit -n 100 f()
# 59.4 µs ± 5.04 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# what's the cost of just iterating over a numpy array?
def g():
    for _ in data: pass
%timeit -n 100 g()
# 738 µs ± 42.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

That is, the pure numpy summation is ~160&times; faster than using the Python `sum`, and is even ~7&times; faster than just iterating over a list. Why?

- `array.sum()`: simple memory layout: get access to the underlying data, and then loop across the array all in C (with full static types, no function calls, zero allocations, and even doing multiple additions in parallel with SIMD)
- `sum(array)` (and `for _ in data`): create a python iterate object for `array`, and for each element, call a python function to get the next element (likely needs a hash table look-up to find the function), which requires allocating a new python object value to hold the C integer, filling in the value, and returning that; then (for `sum(array)`), add that to the accumulator inside `sum` by calling other Python functions (`int.__add__`), which likely involves various dynamic overheads like hash table lookups
- `for _ in L`: create a python iterator object for `L`, and for each element, call a python function to get the next element (with the hash table lookups). This form doesn't have to allocate new objects for the return value (since the `None` values already exist in the list), nor do anything with them, so it's focusing on just the cost of iterating in Python.

There's also sorts of other Python overheads too, like just the pervasive cost of having to interpret Python byte code, rather than executing CPU instructions directly.